File: LeetCode/C++_Solutions/191_Number-Of-1-Bits.cpp
#include <iostream>
using namespace std;

// it processes only the 1 bits, which drastically reduce the number of iterations for numbers with sparse set bits.
class Solution {
public:
    int hammingWeight(int n) {
        int setBitCount = 0;
        while (n) {
            n &= (n - 1); // Removes the lowest set bit
            ++setBitCount; 
        }

        return setBitCount;
    }
};

File: LeetCode/Java_Solutions/191_Number-Of-1-Bits.java
// efficient solution
class Solution {
    public int hammingWeight(int n) {
        int setBitCount = 0;
        while(n != 0){
            n &= (n-1);
            ++setBitCount;
        }
        return setBitCount;
    }
}

// this one is good and creative
class Solution {
    public int hammingWeight(int n) {
        int setBitCount = 0;
        while(n != 0){
            setBitCount += (n%2);
            n /= 2;
        }
        return setBitCount;
    }
}